### ISSUE TITLE
E-graph backed context morphisms

### ISSUE THREAD

**Issue by olynch** [AlgebraicJulia/Gatlab.jl#46]
A morphism between context A and context B should be an e-graph generated by the variables in context A, along with a choice of e-class for each variable in context B. Then composition of f and g should happen via disjoint union of the two e-graphs and then merging of the e-classes for the variables in B in g with the corresponding terms in f.

This representation is a good intermediate step for optimizing the compilation of the morphism as a whole, as each expression is only computed once.

**Comment by jpfairbanks** [comment]
I don't think that Owen is planning to use Metatheory directly, but stay tuned for PRs about egraphs as your expertise from implementing Metatheory is certainly welcome as we implement our own version.

I think that there are some special needs for Gatlab because of the dependent typing in GATs that prevents us from directly using MT. Or at least that is my recollection from our previous discussions about MT.

**Comment by 0x0f0f0f** [comment]
I have released 2.0 - things did change a lot since last time we chatted and MT is a bit more solid and most importantly extendable, and I have some plans for improvements!

**Comment by 0x0f0f0f** [comment]
https://juliasymbolics.github.io/Metatheory.jl/dev/tutorials/custom_types/

Please see this. Are you guys using pushout rewriting or double pushout?

**Comment by olynch** [comment]
There's some stuff that's pretty special to GATs; I use the e-graph to perform type inference. That *might* be possible with metatheory metadata, but I'm not sure. E-Graphs are a pretty core part of Gatlab, and having everything be concretized to Gatlab types makes things a lot simpler. I'm happy to talk about using metatheory, but currently I have no plans to switch. On the other hand, if @0x0f0f0f had some dedicated funding/time carved out for supporting Gatlab, I'd more seriously consider switching.

**Comment by 0x0f0f0f** [comment]
Interested in looking for some funding and time for the project.

**Comment by 0x0f0f0f** [comment]
Rewriting a performant e-graphs system from scratch is a big task though. I'm mostly interested in seeing Gatlab as a concrete application for Metatheory.jl. I'm planning to improve and solidify the metadata interface, which already works nicely in 2.0.
